express.js

```diff
@@ -1,63 +1,108 @@
+    }
+
+    res.json(user);
 });
 
+
+// TODO user can have email which is unique
+// TODO userId must be UUID
+// user add operation needs to recieve metadata and then spread
+
+// get operation with id and email needs to be o(1)
+
 app.post('/users', (req, res) => {
-    let info = readUsers();
-    console.log(info);
-    const { name, surname } = req.body;
+    let users = readUsers();
+    const { name, surname, email, meta } = req.body;
+    
+    if (Object.values(users).some(u => u.email === email)) {
+        return res.status(404).json({ letter: "Email must be unique" });
+    }
+
+    let id = uuidv4();
+    const newUser = { ...meta, name, surname, email };
+    users[id] = newUser;
     
-    let users = info.users;
-    let newId = users.length > 0 ? Number(users[users.length - 1].id) + 1 : 1;
-    const newUser = { id: newId, name, surname };
-    users.push(newUser);
-    writeUsers({ users });
+    writeUsers(users);
     res.status(201).json(newUser);
 });
 
 
 app.delete('/users/:id', (req, res) => {
-   let info = readUsers();
-   let users = info.users;
-   const id = parseInt(req.params.id);
-   const index = users.findIndex(u => u.id == id);
+   let users = readUsers();
+   const id = req.params.id;
 
-   if (index == -1) {
-        return res.status(404).json("User not found");
+   if (!users[id]) {
+        return res.status(404).json({ letter: "User not found" });
    }
 
-   const deletedUser = users.splice(index, 1)[0];
-   writeUsers({ users });
+   const deletedUser = users[id];
+   delete users[id];
+
+   writeUsers(users);
    res.json({ letter: "User deleted", user: deletedUser });
 });
 
-app.put('/users/:id', (req, res) => {
-   let info = readUsers();
-   const id = parseInt(req.params.id);
-   const { name, surname } = req.body;
 
-   let users = info.users;
-   const user = users.find(u => id == u.id);
+   // TODO add patch
+app.patch('/users/:id', (req, res) => {
+    let users = readUsers();
+    const id = req.params.id;
 
-   if (!user.name || !user.surname) {
-    return res.status(404).json("User not found");
+    if (!users[id]) {
+        return res.status(404).json({ message: "User not found"});;
+    }
+
+    const update = req.body;
+
+    if (update.email && Object.values(users).some(u => u.email === update.email && u.id !== id)) {
+        return res.status(404).json({ message: "Email must be unique" }); 
+    }
+
+    users[id] = { ...users[id], ...update };
+    writeUsers(users);
+
+    res.json(users[id]);
+});
+
+
+
+app.put('/users/:id', (req, res) => {
+   let users = readUsers();
+   const id = req.params.id;
+
+   if (!users[id]) {
+        return res.status(404).json({ message: "User not found"});;
    }
 
-   user.name = name === undefined ? user.name : name;
-   user.surname = surname === undefined ? user.surname : surname;
+   const { name, surname, email, meta } = req.body;
+
+   if (email && Object.values(users).some(u => u.email === email && u.id !== id)) {
```

Comment:
same issue, every time you have the exact same code repeating means something is needs to be optimised